pkg/filter/filter.go

```diff
@@ -179,6 +201,24 @@ func Predicate() predicate.Predicate {
 			e.ObjectOld.GetGeneration()
 			e.ObjectOld.GetOwnerReferences()
 
+			obj := e.ObjectNew
+
+			// Required for the case when pods are deleted due to OS upgrade
```

Comment:
I am not sure if its desirable to introduce such specific behavior here. If we only check for daemonsets we could be leaving out other potential ways of deploying driver containers, such as pods and/or deployments. Same happens with the kernel affine label, we might have other updates that dont fall into that category and we would let them go through.

Try this and see if it works:
```suggestion
			if !Owned(obj) {
				return false
			}

			if e.ObjectOld.GetGeneration() == e.ObjectNew.GetGeneration() &&
				e.ObjectOld.GetResourceVersion() == e.ObjectNew.GetResourceVersion() {
				return false
			}

			if IsSpecialResource(obj) {
				log.Info(Mode+" IsSpecialResource GenerationChanged",
					"Name", obj.GetName(), "Type", reflect.TypeOf(obj).String())
				return true
			}

			log.Info(Mode+" Owned GenerationChanged",
				"Name", obj.GetName(), "Type", reflect.TypeOf(obj).String())

			if reflect.TypeOf(obj).String() == "*v1.DaemonSet" {
				err := lifecycle.UpdateDaemonSetPods(obj)
				warn.OnError(err)
			}

			return true
```

Reply:
For pod it should not be a problem because SRO owns it directly. In case of deployment it will spawn the pod on some other node if the pod on one node is deleted. Kernel affine Daemonset and OS upgrade is the only combination where we can face issue. Non kernel affine daemonset itself will handle pod creation without the SRO reconciler running. That's why restricted the check to kernel affine SRO owned daemonset. With this change it would trigger reconciler multiple times even when the deployment or daemonset pod readiness change. 
Also, UpdateDaemonSetPods should only be called when the generation changes in case of resourceVersion change it will falsely add entries of all pods in configmap and keep waiting for them to be updated.

Reply:
Let's see all the different situations to see if I am mistaken. Any of the following objects will be directly owned by SRO if specified like that in the recipe. Keep in mind that the current update function ignores `status` changes and only looks for `spec` ones. Let's suppose that all of them use the kernel affine label as well.
Pod:
A stand-alone pod will be running only 1 replica in some node. If the kernel is updated there will be a reboot and the pod will be scheduled in a node where the kernel is affine. This will happen until there are no nodes left with the same kernel version. Since the reschedules are changing the resourceVersion but not the generation, we dont reconcile and the nodeSelector is not updated.

Deployment:
Similar to the previous one, the only difference is that we can have more than one replica. The end situation is the same though, it will keep rescheduling all the pods to kernel affine nodes until there are no more left. Since the reschedules are changing the resourceVersion but not the generation, we dont reconcile and the nodeSelector is not updated.

Daemonset:
Your PR got this one right, no comments here because it does what it needs to do.

In all cases the SRO pod itself will get rescheduled when its node restarts, but there is no order guarantee so we cant assume it will happen after all the other SRO owned resources have been rebooted as well.

Am I right or did I miss something?

Reply:
Yes you are right, we have 3 resources to take care of. For Pod and Deployment we do not have node dependency. They can be scheduled on node that meets the criteria. When the nodes are updating SRO pod will restart when the node on which SRO is running is rebooted, that will make a call to reconciler. At this point it will update the variables so that the deployment pods and pod start scheduling on the nodes that have updated OS.

There is no harm even if we call reconciler every time when the status is updated. It will just create additional overhead for the reconcile cycles even when nothing changed. We thought of multiple cases and think pod and deployment will not have any issue. If we are missing any case, we can relax the check for resourceVersion and generation mismatch to all SRO owned kernel affine objects instead of daemonset only.

Reply:
> Yes you are right, we have 3 resources to take care of. For Pod and Deployment we do not have node dependency. They can be scheduled on node that meets the criteria. When the nodes are updating SRO pod will restart when the node on which SRO is running is rebooted, that will make a call to reconciler. At this point it will update the variables so that the deployment pods and pod start scheduling on the nodes that have updated OS.

This is true, but there is one caveat. In large clusters the workloads might stop working way before the node in which SRO is running is rebooted, delaying reconcile. This could be prevented by preemptively reconciling.

> There is no harm even if we call reconciler every time when the status is updated. It will just create additional overhead for the reconcile cycles even when nothing changed. We thought of multiple cases and think pod and deployment will not have any issue. If we are missing any case, we can relax the check for resourceVersion and generation mismatch to all SRO owned kernel affine objects instead of daemonset only.

The overhead is the only drawback of this, but I think its preferrable to have a useless reconcile cycle than stopping workloads until reboots happen. That is why I suggested a more generic approach, not only checking for daemonsets.

Reply:
Okay will relax the check to check SRO owned kernel affine objects instead of SRO owned kernel affine daemonsets.

Reply:
I think now its as flexible as it needs to be, thanks!